### DISCUSSION TITLE
Project Goals, Coding Standards, etc.

### DISCUSSION THREAD

**Discussion by henrygab** [forkineye/ESPixelStick#520] [General]
**_NOTE: This contains a summary, based on discussions with frequent contributor MartinMueller2003.  There is no indication that Shelby has reviewed or approved these (yet).  Thus, they are simply a best-effort attempt to help folks create PRs that already conform to current practices, and thus make it easier for Shelby's to review / accept them._**

I'm using this first post for a summary, which will update as needed with clarifications over time:

1. This project is tied to Arduino framework and libraries.

2. This project doesn't change default C++ version in Arduino.
    a. Thus, currently Gnu++11 defines the available C++ features.

3. Restrict code to simple / hobbyist-accessible C++ feature.
    a. Avoid templates; Simplest usage of templates **_may_** be OK
    b. Avoid metaprogramming
    c. Avoid adding namespaces
    d. Do **_NOT_** rely on throwing or handling C++ exceptions

4. Avoid heap allocations during "runtime".  Martin put it well:
    > Allocate once at start up and then never allocate or free again. - Free in the destructor and that's it. You will find that the input and output modules only new / delete at the time the inputs or outputs are changing due to configuration changes. Since we free before we new, I do not worry too much about a failed allocation. - All other buffers and memory are allocated at compile time or boot time.

5. IRAM limitations (no flash access in ISRs) impact preprocessor use
    a. Generally, scoped enumerations are encouraged for multiple related options
    b. **_Individual_** global constants should be declared using `#define` (safe to use in ISR)
    c. Consider if **_global_** constants should be in flash (use `const` global, but no ISR use), and if not, use `#define`
    d. Macros are only convertible to inline functions if not used by any ISR, or are marked with `IRAM_ATTR`

6. Stylistic preferences
    a. Per MartinMueller2003, prefer initialization of class fields to occur [in the header](https://github.com/forkineye/ESPixelStick/pull/523#discussion_r862292838), rather than base-class-style initialization in header.

**Comment by MartinMueller2003** [comment]
1. Compiler flags enable exceptions (`-fexception`) ... Is this intentional? 
   
   - Over time I have found that exceptions in real time code are not very useful. Therefore, I have not used them in the V4 code.


2. Is there an explicit goal to build at a minimum version of C++?

   
   - We are aligned with the compiler approved / implemented / tested by the Espressif and Arduino teams. When they move forward, so will we. The compiler is installed by the Platformio Arduino package and I prefer not to override it and find about which libs have interesting issues. There are a few V14 features I would like to use, but 11 is perfectly serviceable.
3. Is the default rule that functions should run without dynamic memory allocation?

   
   - There is no hard and fast rule but you will only find malloc in the default Document constructor for Arduino documents being placed into PSRAM. The WiFi and Ethernet libs may be allocating and deleting memory as part of the DMA channel processing but that is invisible to us.
   - My general policy on using heap in a real time environment is: Allocate once at start up and then never allocate or free again.
   - Free in the destructor and that's it. You will find that the input and output modules only new / delete at the time the inputs or outputs are changing due to configuration changes. Since we free before we new, I do not worry too much about a failed allocation.
   - All other buffers and memory are allocated at compile time or boot time.
4. Conversely, is the default rule that dynamic memory allocation is fine?

   
   - See above. IMO Dynamic allocation has no place in a real time system that depends on deterministic processing time.
5. How to prevent exceptions from ruining a show?

   
   - The code has its roots in Arduino. Arduino has rewritten the stl functions to work in a small memory footprint model. You can override and use them, but since the bulk of the code was written in the Arduino model, there is almost no stl code. I would love to use vectors instead of arrays but we have what we have. And no we cannot move away from Arduino framework at this time. Most of the TCP and UDP libes are Arduino libs and I do not have the time to look for STL compliant versions of that functionality.
   - I avoid using templates (Yes there are some in the code but they can become a nightmare for hobbyists to understand).
6. Any interest in reducing reliance on pre-processor, including `#define` constants and function-like macros...?

The code uses preprocessor defines instead of const vars and because they allow me to make it obvious what is being done. There are a few macros such as Enqueue and fifolength that are carry overs from the V3 days and they could become an inline function. The IRAM rules do make it difficult to use things like templates to do this functionality. For instance, you cannot access a const var from the ISRs etc. 
To be honest, most true C++ code is very jarring to the rest of us and is not hobbiest friendly. The code is written in a way to be easy to understand for moderate hoppy developers and therefore easy to modify. Deep C++ can make that very difficult.

| 
| 
| 
|  |  |

 |

 |
| 
|  | 
Shall I prefer constants over defines?

In C, shall I prefer constants over defines? I've reading a lot of code lately, and all of the examples make hea...
 |

 |

 |




Yes, I might volunteer to do this, as each change would be a small, bite-sized change.

Example techniques might include:
   
   - Use of C++11 namespaces (avoid collisions, global namespace pollution, etc.)
   - Use of constexpr functions instead of macros
   - When absolutely necessary to leave, having a comment indicating the limitation causing its use.

I am intentionally avoiding the use of namespaces and do not have any name collision issues. I find the use of name spaces to be one of those mechanisms that would make the code difficult for non C++ programmers to understand.
We could convert the macros to constexpr as long as none of them show up in an ISR.
Martin
    On Saturday, April 23, 2022, 06:22:58 PM EDT, Henry Gabryjelski ***@***.***> wrote:  
 
 


Shelby, Martin --

First, I think the direction taken with the v4 code is good.
   
   - The layering and class object names make sense, and abstract reasonably.
   - The use of newer-style for-loops is welcome and refreshing.
   - The use of const and references -- "Thank You!"

At the same time, before I start digging into the code extensively, or providing feedback, I find some things are really good to know.

Can you help me understand some goals for this code?

Here's a few questions (with collapsed-by-default comments):
 1. Compiler flags enable exceptions (`-fexception`) ... Is this intentional?

* Pros and cons to either choice ... but an explicit choice is necessary
* Is think Arduino 2.x has chosen to enable exceptions
* But, could use help understanding expectations if any exception is ever actually thrown...
 2. Is there an explicit goal to build at a minimum version of C++?

   
   - Based on platformio "verbose build", currently many files are built with -std=gnu++11
   - Why not C++14 / gnu++14?
 3. Is the default rule that functions should run without dynamic memory allocation?

   
   - Even if exceptions are enabled, this can improve stability and reduce error paths.
   - Only a default, because exceptions nearly always exists...      
      - e.g., initialization and setup, such as using Arduino's begin()/end() conventions
      - e.g., when effectively reconfiguring the hardware, such as applying a new configuration
      - e.g., when the failure paths are tested, and effects of failing midway are understood

 4. Conversely, is the default rule that dynamic memory allocation is fine?

   
   - Exceptions obviously exist .. such as ISRs and the like
   - If yes, is there a desire to avoid dynamic memory allocation more often in the future?      
      - This is worth a separate discussion if this ends up "yes"...

 5. How to prevent exceptions from ruining a show?

   
   - I noticed an explicit avoidance of the use of STL ... which is a good step, since I don't think there's any way to restrict when STL allocates memory
   - Is STL usage restricted to compile-time aspects (templates, template metaprogramming helpers, etc.)?
   - 
 6. Any interest in reducing reliance on pre-processor, including `#define` constants and function-like macros...?


The code is entirely C++. For C, there are reasons to use #define (or better, to make those enum values). But for C++, it's quite jarring ... especially when this is not for a library header file defining items that would be used from clients written in C.

Yes, I might volunteer to do this, as each change would be a small, bite-sized change.

Example techniques might include:
   
   - Use of C++11 namespaces (avoid collisions, global namespace pollution, etc.)
   - Use of constexpr functions instead of macros
   - When absolutely necessary to leave, having a comment indicating the limitation causing its use.

Your (or Shelby's) answers will help me understand the code, and suggest changes that meet the goals of the project.

I'm a big fan ... I love the stability of the v3.x firmware, and look forward to making v4 worthy of the ESPixelStick name!

**Reply by henrygab** [reply]
The following is the original post, moved here for posterity.

Shelby, Martin -- 

First, I think the direction taken with the v4 code is good.
* The layering and class object names make sense, and abstract reasonably.
* The use of newer-style for-loops is welcome and refreshing.
* The use of `const` and references -- "Thank You!"

At the same time, before I start digging into the code extensively, or providing feedback, I find some things are really good to know.   

**_Can you help me understand some goals for this code?_**

Here's a few questions (with collapsed-by-default comments):
<HR/><details><summary>
1.  Compiler flags enable exceptions (`-fexception`) ... Is this intentional?
</summary><P/>

    * Pros and cons to either choice ... but an explicit choice is necessary
    * Is think Arduino 2.x has chosen to enable exceptions
    * But, could use help understanding expectations if any exception is ever actually thrown...

</details><HR/><details><summary>
2. Is there an explicit goal to build at a minimum version of C++?
</summary><P/>

* Based on platformio "verbose build", currently many files are built with `-std=gnu++11`
* Why not C++14 / gnu++14?
</details><HR/><details><summary>
3. Is the default rule that functions should run without dynamic memory allocation?
</summary><P/>

* Even if exceptions are enabled, this can improve stability and reduce error paths.
* Only a default, because rule-exceptions nearly always exists...
  * e.g., initialization and setup, such as using Arduino's `begin()`/`end()` conventions
  * e.g., when effectively reconfiguring the hardware, such as applying a new configuration
  * e.g., when the failure paths are tested, and effects of failing midway are understood
</details><HR/><details><summary>
4. Conversely, is the default rule that dynamic memory allocation is fine?
</summary><P/>

* Rule-exceptions obviously exist .. such as ISRs and the like
* If yes, is there a desire to avoid dynamic memory allocation more often in the future?
  *  This is worth a separate discussion if this ends up "yes"...
</details><HR/><details><summary>
5. How to prevent exceptions from ruining a show?
</summary><P/>

* I noticed an explicit avoidance of the use of STL ... which is a good step, since I don't think there's any way to restrict when STL allocates memory
* Is STL usage restricted to compile-time aspects (templates, template metaprogramming helpers, etc.)?

</details><HR/><details><summary>
6. Any interest in reducing reliance on pre-processor, including `#define` constants and function-like macros...?
</summary><P/>

The code is entirely C++.  For C, there are [reasons](https://stackoverflow.com/questions/2308194/shall-i-prefer-constants-over-defines/2308364#2308364) to use `#define` (or better, to make those `enum` values).  But for C++, it's quite jarring ... especially when this is not for a library header file defining items that would be used from clients written in C.

Yes, I might volunteer to do this, as each change would be a small, bite-sized change.

Example techniques might include:
1. Use of C++11 namespaces (avoid collisions, global namespace pollution, etc.)
2. Use of `constexpr` functions instead of macros
3. When absolutely necessary to leave, having a comment indicating the limitation causing its use.
</details><HR/>

Your (or Shelby's) answers will help me understand the code, and suggest changes that meet the goals of the project.

I'm a big fan ... I love the stability of the v3.x firmware, and look forward to making v4 worthy of the ESPixelStick name!

**Comment by henrygab** [comment]
Did I properly split hairs between 5a and 5b above?  (scoped enumerations when multiple related options, vs. `#define` for individual constant values)

Anything else I should edit/update from the summary?

For example, are there any (automated) code formatting options that should be applied, prior to submitting a pull request?

This is an excellent set of guidelines... THANK YOU!  I will strive to meet them, for any PRs I submit.

**Comment by MartinMueller2003** [comment]
It is not that const int is discouraged, you need to be careful that const vars are not used in an ISR. That means knowing when they are being used which can be a bit confusing and potentially a pitfall. 

The rest looks fine.
Martin
    On Sunday, April 24, 2022, 04:55:28 PM EDT, Henry Gabryjelski ***@***.***> wrote:  
 
 


Did I properly split hairs between 5a and 5b above? (scoped enumerations when multiple related options, vs. #define for individual constant values)

Anything else I should edit/update from the summary?

For example, are there any (automated) code formatting options that should be applied, prior to submitting a pull request?

This is an excellent set of guidelines... THANK YOU! I will strive to meet them, for any PRs I submit.

**Comment by MartinMueller2003** [comment]
FYI: I am in the process of restructuring the UART output functionality and then I will take a pass at normalizing the APIs in a way that simplifies the tuning. That is one of the reasons I wish we had vectors. I could put tables in PROGMEM and simply pass a reference to the table for the desired output characteristics to the driver. That significantly simplifies the setup of the pixel protocols.
Martin
    On Sunday, April 24, 2022, 09:31:56 AM EDT, Martin Mueller ***@***.***> wrote:  
 
  1. Compiler flags enable exceptions (`-fexception`) ... Is this intentional? 
   
   - Over time I have found that exceptions in real time code are not very useful. Therefore, I have not used them in the V4 code.


2. Is there an explicit goal to build at a minimum version of C++?

   
   - We are aligned with the compiler approved / implemented / tested by the Espressif and Arduino teams. When they move forward, so will we. The compiler is installed by the Platformio Arduino package and I prefer not to override it and find about which libs have interesting issues. There are a few V14 features I would like to use, but 11 is perfectly serviceable.
3. Is the default rule that functions should run without dynamic memory allocation?

   
   - There is no hard and fast rule but you will only find malloc in the default Document constructor for Arduino documents being placed into PSRAM. The WiFi and Ethernet libs may be allocating and deleting memory as part of the DMA channel processing but that is invisible to us.
   - My general policy on using heap in a real time environment is: Allocate once at start up and then never allocate or free again.
   - Free in the destructor and that's it. You will find that the input and output modules only new / delete at the time the inputs or outputs are changing due to configuration changes. Since we free before we new, I do not worry too much about a failed allocation.
   - All other buffers and memory are allocated at compile time or boot time.
4. Conversely, is the default rule that dynamic memory allocation is fine?

   
   - See above. IMO Dynamic allocation has no place in a real time system that depends on deterministic processing time.
5. How to prevent exceptions from ruining a show?

   
   - The code has its roots in Arduino. Arduino has rewritten the stl functions to work in a small memory footprint model. You can override and use them, but since the bulk of the code was written in the Arduino model, there is almost no stl code. I would love to use vectors instead of arrays but we have what we have. And no we cannot move away from Arduino framework at this time. Most of the TCP and UDP libes are Arduino libs and I do not have the time to look for STL compliant versions of that functionality.
   - I avoid using templates (Yes there are some in the code but they can become a nightmare for hobbyists to understand).
6. Any interest in reducing reliance on pre-processor, including `#define` constants and function-like macros...?

The code uses preprocessor defines instead of const vars and because they allow me to make it obvious what is being done. There are a few macros such as Enqueue and fifolength that are carry overs from the V3 days and they could become an inline function. The IRAM rules do make it difficult to use things like templates to do this functionality. For instance, you cannot access a const var from the ISRs etc. 
To be honest, most true C++ code is very jarring to the rest of us and is not hobbiest friendly. The code is written in a way to be easy to understand for moderate hoppy developers and therefore easy to modify. Deep C++ can make that very difficult.

| 
| 
| 
|  |  |

 |

 |
| 
|  | 
Shall I prefer constants over defines?

In C, shall I prefer constants over defines? I've reading a lot of code lately, and all of the examples make hea...
 |

 |

 |




Yes, I might volunteer to do this, as each change would be a small, bite-sized change.

Example techniques might include:
   
   - Use of C++11 namespaces (avoid collisions, global namespace pollution, etc.)
   - Use of constexpr functions instead of macros
   - When absolutely necessary to leave, having a comment indicating the limitation causing its use.

I am intentionally avoiding the use of namespaces and do not have any name collision issues. I find the use of name spaces to be one of those mechanisms that would make the code difficult for non C++ programmers to understand.
We could convert the macros to constexpr as long as none of them show up in an ISR.
Martin
    On Saturday, April 23, 2022, 06:22:58 PM EDT, Henry Gabryjelski ***@***.***> wrote:  
 
 


Shelby, Martin --

First, I think the direction taken with the v4 code is good.
   
   - The layering and class object names make sense, and abstract reasonably.
   - The use of newer-style for-loops is welcome and refreshing.
   - The use of const and references -- "Thank You!"

At the same time, before I start digging into the code extensively, or providing feedback, I find some things are really good to know.

Can you help me understand some goals for this code?

Here's a few questions (with collapsed-by-default comments):
 1. Compiler flags enable exceptions (`-fexception`) ... Is this intentional?

* Pros and cons to either choice ... but an explicit choice is necessary
* Is think Arduino 2.x has chosen to enable exceptions
* But, could use help understanding expectations if any exception is ever actually thrown...
 2. Is there an explicit goal to build at a minimum version of C++?

   
   - Based on platformio "verbose build", currently many files are built with -std=gnu++11
   - Why not C++14 / gnu++14?
 3. Is the default rule that functions should run without dynamic memory allocation?

   
   - Even if exceptions are enabled, this can improve stability and reduce error paths.
   - Only a default, because exceptions nearly always exists...      
      - e.g., initialization and setup, such as using Arduino's begin()/end() conventions
      - e.g., when effectively reconfiguring the hardware, such as applying a new configuration
      - e.g., when the failure paths are tested, and effects of failing midway are understood

 4. Conversely, is the default rule that dynamic memory allocation is fine?

   
   - Exceptions obviously exist .. such as ISRs and the like
   - If yes, is there a desire to avoid dynamic memory allocation more often in the future?      
      - This is worth a separate discussion if this ends up "yes"...

 5. How to prevent exceptions from ruining a show?

   
   - I noticed an explicit avoidance of the use of STL ... which is a good step, since I don't think there's any way to restrict when STL allocates memory
   - Is STL usage restricted to compile-time aspects (templates, template metaprogramming helpers, etc.)?
   - 
 6. Any interest in reducing reliance on pre-processor, including `#define` constants and function-like macros...?


The code is entirely C++. For C, there are reasons to use #define (or better, to make those enum values). But for C++, it's quite jarring ... especially when this is not for a library header file defining items that would be used from clients written in C.

Yes, I might volunteer to do this, as each change would be a small, bite-sized change.

Example techniques might include:
   
   - Use of C++11 namespaces (avoid collisions, global namespace pollution, etc.)
   - Use of constexpr functions instead of macros
   - When absolutely necessary to leave, having a comment indicating the limitation causing its use.

Your (or Shelby's) answers will help me understand the code, and suggest changes that meet the goals of the project.

I'm a big fan ... I love the stability of the v3.x firmware, and look forward to making v4 worthy of the ESPixelStick name!

**Reply by henrygab** [reply]
If you can edit to remove most of the replied-to text ... leaving in many pages makes it difficult to see what part of the conversation you're responding to....
